src/wallet/coinselection.cpp

```diff
@@ -188,16 +200,51 @@ util::Result<SelectionResult> SelectCoinsSRD(const std::vector<OutputGroup>& utx
     Shuffle(indexes.begin(), indexes.end(), rng);
 
     CAmount selected_eff_value = 0;
+    int weight = 0;
+    bool max_tx_size_exceeded = false;
     for (const size_t i : indexes) {
         const OutputGroup& group = utxo_pool.at(i);
         Assume(group.GetSelectionAmount() > 0);
+
+        // If the selection weight exceeds the maximum allowed size, remove the least valuable inputs until we
+        // can insert the new group (while the sum of them are less than the new group amount).
+        weight += group.m_weight;
+        if (weight > max_weight) {
+            max_tx_size_exceeded = true; // mark it in case we don't find any useful result.
+
+            // if the least valuable input is greater, discard the new group.
+            CAmount new_group_amount = group.GetSelectionAmount();
+            if (new_group_amount <= heap.top().GetSelectionAmount()) {
+                weight -= group.m_weight;
+                continue;
+            }
+
+            // Keep popping up inputs while we exceed the max weight and the
+            // arriving 'group' amount is less than the removed inputs.
+            CAmount least_valuable_amounts = 0;
+            do {
+                const OutputGroup& to_remove_group = heap.top();
+                CAmount min_group_amount = to_remove_group.GetSelectionAmount();
+                least_valuable_amounts += min_group_amount;
+                selected_eff_value -= min_group_amount;
+                weight -= to_remove_group.m_weight;
+                heap.pop();
+            } while (!heap.empty() && least_valuable_amounts < new_group_amount && weight > max_weight);
+        }
+
+        // Add group to selection
         selected_eff_value += group.GetSelectionAmount();
-        result.AddInput(group);
+        heap.push(group);
```

Comment:
It's simpler if you first add the new group to the heap and then use the heap to get below weight:

```suggestion
          
        // Add group to selection
        heap.push(group);
        selected_eff_value += group.GetSelectionAmount();
        weight += group.m_weight;
        
        // If the selection weight exceeds the maximum allowed size, remove the least valuable inputs until we
        // can insert the new group (while the sum of them are less than the new group amount).
        if (weight > max_weight) { 
            max_tx_size_exceeded = true; // mark it in case we don't find any useful result.

            do {
                const OutputGroup& to_remove_group = heap.top();
                CAmount min_group_amount = to_remove_group.GetSelectionAmount();
                selected_eff_value -= min_group_amount;
                weight -= to_remove_group.m_weight;
                heap.pop();
            } while (!heap.empty() && weight > max_weight);
        }
```

Reply:
Nice 👌🏼, applied changes and added your co-authorship. Thanks :).